File: 94_Binary_Tree_Inorder_Traversal.java
class Solution {

    public List<Integer> inorderTraversal(TreeNode root) {

        /* Sanity check */
        if(root == null)
            return new ArrayList<>();

        Stack<TreeNode> s = new Stack<>();
        ArrayList<Integer> result = new ArrayList<>();
        TreeNode n = root;

        while(true) {
            /* Push all left child to stack */
            while(n != null) {
                s.push(n);
                n = n.left;
            }

            /* Termination */
            if(s.empty())
                return result;

            n = s.pop();
            result.add(Integer.valueOf(n.val));
            n = n.right;
        }
    }
}

File: 94_Binary_Tree_Inorder_Traversal.cpp
class Solution {
 public:
  vector<int> inorderTraversal(TreeNode* root) {
    stack<TreeNode*> address;
    vector<int> result;
    TreeNode* cur = root;
    if (!root) return result;

    while (1) {
      while (cur) {  // to the leftmost
        address.push(cur);
        cur = cur->left;
      }
      if (address.empty()) break;
      cur = address.top();
      address.pop();
      result.push_back(cur->val);
      cur = cur->right;
    }
    return result;
  }
};